File: app/src/main/java/com/noobfitness/noobfitness/legacy/MainActivity.java
package com.noobfitness.noobfitness.legacy;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.common.SignInButton;
import com.google.android.material.snackbar.Snackbar;
import com.noobfitness.noobfitness.R;
import com.squareup.picasso.Picasso;

import net.openid.appauth.AuthState;
import net.openid.appauth.AuthorizationException;
import net.openid.appauth.AuthorizationResponse;
import net.openid.appauth.AuthorizationService;
import net.openid.appauth.TokenResponse;

import org.json.JSONException;
import org.json.JSONObject;

import okhttp3.FormBody;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class MainActivity extends Activity {

    private static final String LOG_TAG = "MainActivity";

    AuthState mAuthState;

    Button mMakeApiCall;
    Button mSignOut;
    TextView mGivenName;
    TextView mFamilyName;
    TextView mFullName;
    ImageView mProfileView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.legacy_activity_main);

        mMakeApiCall = findViewById(R.id.makeApiCall);
        mSignOut = findViewById(R.id.signOut);
        mGivenName = findViewById(R.id.givenName);
        mFamilyName = findViewById(R.id.familyName);
        mFullName = findViewById(R.id.fullName);
        mProfileView = findViewById(R.id.profileImage);

        handleAuthorizationResponse(getIntent());
        enablePostAuthorizationFlows();

        LinearLayout mainLinearLayout = findViewById(R.id.activity_main);

        Button fourDayDefault = new Button(this);
        fourDayDefault.setText("4-day default routine");
        fourDayDefault.setGravity(Gravity.LEFT);
        fourDayDefault.setPadding(96, 96, 96, 96);
        fourDayDefault.setTypeface(Typeface.create("sans-serif-light", Typeface.NORMAL));
        fourDayDefault.setTextSize(24);
        fourDayDefault.setAllCaps(false);
        fourDayDefault.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent i = new Intent(v.getContext(), RoutinesActivity.class);
                startActivity(i);
            }
        });
        mainLinearLayout.addView(fourDayDefault);

        TextView imageCredit = new TextView(this);
        imageCredit.setText("Image source: Everkinetic.com (CC BY-SA 3.0)");
        mainLinearLayout.addView(imageCredit);

    }

    /**
     * Exchanges the code, for the {@link TokenResponse}.
     *
     * @param intent represents the {@link Intent} from the Custom Tabs or the System Browser.
     */
    private void handleAuthorizationResponse(@NonNull Intent intent) {

        // code from the step 'Handle the Authorization Response' goes here.
        AuthorizationResponse response = AuthorizationResponse.fromIntent(intent);
        AuthorizationException error = AuthorizationException.fromIntent(intent);
        final AuthState authState = new AuthState(response, error);

        if (response != null) {
            Log.i(LOG_TAG, String.format("Handled Authorization Response %s ", authState.jsonSerializeString()));
            AuthorizationService service = new AuthorizationService(this);
            service.performTokenRequest(response.createTokenExchangeRequest(), new AuthorizationService.TokenResponseCallback() {
                @Override
                public void onTokenRequestCompleted(@Nullable TokenResponse tokenResponse, @Nullable AuthorizationException exception) {
                    if (exception != null) {
                        Log.w(LOG_TAG, "Token Exchange failed", exception);
                    } else {
                        if (tokenResponse != null) {
                            authState.update(tokenResponse, exception);
                            persistAuthState(authState);
                            Log.i(LOG_TAG, String.format("Token Response [ Access Token: %s, ID Token: %s ]", tokenResponse.accessToken, tokenResponse.idToken));
                        }
                    }
                }
            });
        }
    }

    private static final String SHARED_PREFERENCES_NAME = "AuthStatePreference";
    private static final String AUTH_STATE = "AUTH_STATE";

    private void persistAuthState(@NonNull AuthState authState) {
        getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE).edit()
                .putString(AUTH_STATE, authState.jsonSerializeString())
                .commit();
        enablePostAuthorizationFlows();
    }

    private void enablePostAuthorizationFlows() {
        mAuthState = restoreAuthState();
        if (mAuthState != null && mAuthState.isAuthorized()) {
            if (mMakeApiCall.getVisibility() == View.GONE) {
                mMakeApiCall.setVisibility(View.VISIBLE);
                mMakeApiCall.setOnClickListener(new MainActivity.MakeApiCallListener(this, mAuthState, new AuthorizationService(this)));
            }
            if (mSignOut.getVisibility() == View.GONE) {
                mSignOut.setVisibility(View.VISIBLE);
                mSignOut.setOnClickListener(new MainActivity.SignOutListener(this));
            }
        } else {
            mMakeApiCall.setVisibility(View.GONE);
            mSignOut.setVisibility(View.GONE);
        }
    }

    private void clearAuthState() {
        getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE)
                .edit()
                .remove(AUTH_STATE)
                .apply();
    }

    @Nullable
    private AuthState restoreAuthState() {
        String jsonString = getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE)
                .getString(AUTH_STATE, null);
        if (!TextUtils.isEmpty(jsonString)) {
            try {
                return AuthState.jsonDeserialize(jsonString);
            } catch (JSONException jsonException) {
                // should never happen
            }
        }
        return null;
    }

    public static class SignOutListener implements Button.OnClickListener {

        private final MainActivity mMainActivity;

        public SignOutListener(@NonNull MainActivity mainActivity) {
            mMainActivity = mainActivity;
        }

        @Override
        public void onClick(View view) {
            mMainActivity.mAuthState = null;
            mMainActivity.clearAuthState();
            mMainActivity.enablePostAuthorizationFlows();
        }
    }

    public static class MakeApiCallListener implements Button.OnClickListener {

        private final MainActivity mMainActivity;
        private AuthState mAuthState;
        private AuthorizationService mAuthorizationService;

        public MakeApiCallListener(@NonNull MainActivity mainActivity, @NonNull AuthState authState, @NonNull AuthorizationService authorizationService) {
            mMainActivity = mainActivity;
            mAuthState = authState;
            mAuthorizationService = authorizationService;
        }

        @Override
        public void onClick(View view) {

            // code from the section 'Making API Calls' goes here
            mAuthState.performActionWithFreshTokens(mAuthorizationService, new AuthState.AuthStateAction() {
                @Override
                public void execute(@Nullable String accessToken, @Nullable String idToken, @Nullable AuthorizationException exception) {
                    new AsyncTask<String, Void, JSONObject>() {
                        @Override
                        protected JSONObject doInBackground(String... tokens) {
                            OkHttpClient client = new OkHttpClient();
                            RequestBody requestBody = new FormBody.Builder()
                                    .add("access_token", tokens[0])
                                    .build();
                            Request request = new Request.Builder()
                                    .post(requestBody)
                                    .url("http://10.0.2.2:5000/api/auth/google")
                                    .build();
                            try {
                                Response response = client.newCall(request).execute();
                                String jsonBody = response.body().string();
                                Log.i(LOG_TAG, String.format("User Info Response %s", jsonBody));
                                return new JSONObject(jsonBody);
                            } catch (Exception exception) {
                                Log.w(LOG_TAG, exception);
                            }
                            return null;
                        }

                        @Override
                        protected void onPostExecute(JSONObject userInfo) {
                            if (userInfo != null) {
                                String fullName = userInfo.optString("username", null);
                                String imageUrl = userInfo.optString("avatar", null);
                                String routines = userInfo.optString("routines", null);
                                if (!TextUtils.isEmpty(imageUrl)) {
                                    Picasso.with(mMainActivity)
                                            .load(imageUrl)
                                            .placeholder(R.drawable.ring_accent)
                                            .into(mMainActivity.mProfileView);
                                }
                                if (!TextUtils.isEmpty(fullName)) {
                                    mMainActivity.mFullName.setText(fullName);
                                }

                                Toast.makeText(mMainActivity, routines, Toast.LENGTH_SHORT).show();
                            }
                        }
                    }.execute(accessToken);
                }
            });
        }
    }
}

File: app/src/main/java/com/noobfitness/noobfitness/legacy/MainActivity.kt
package com.noobfitness.noobfitness.legacy

import android.app.Activity
import android.content.Intent
import android.graphics.Typeface
import android.os.AsyncTask
import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.view.View
import android.widget.*

import com.noobfitness.noobfitness.R
import com.noobfitness.noobfitness.auth.AuthService
import com.noobfitness.noobfitness.auth.UserManager
import com.noobfitness.noobfitness.dagger.InjectedApplication
import com.squareup.picasso.Picasso

import org.json.JSONObject

import javax.inject.Inject

import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request

class MainActivity : Activity() {

    @Inject
    lateinit var authService: AuthService
    @Inject
    lateinit var userManager: UserManager

    private lateinit var makeApiCallButton: Button
    private lateinit var signOutButton: Button
    private lateinit var usernameTextView: TextView
    private lateinit var avatarImageView: ImageView

    private lateinit var fourDayDefault: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        (application as InjectedApplication).appComponent.inject(this)

        setContentView(R.layout.legacy_activity_main)

        makeApiCallButton = findViewById(R.id.makeApiCall)
        signOutButton = findViewById(R.id.signOut)
        usernameTextView = findViewById(R.id.fullName)
        avatarImageView = findViewById(R.id.profileImage)

        enablePostAuthorizationFlows()

        val mainLinearLayout = findViewById<LinearLayout>(R.id.activity_main)

        fourDayDefault = Button(this)
        fourDayDefault.text = userManager.get()?.routines
        fourDayDefault.gravity = Gravity.LEFT
        fourDayDefault.setPadding(96, 96, 96, 96)
        fourDayDefault.typeface = Typeface.create("sans-serif-light", Typeface.NORMAL)
        fourDayDefault.textSize = 24f
        fourDayDefault.isAllCaps = false
        fourDayDefault.setOnClickListener { v ->
            val i = Intent(v.context, RoutinesActivity::class.java)
            startActivity(i)
        }
        mainLinearLayout.addView(fourDayDefault)
    }

    private fun enablePostAuthorizationFlows() {
        if (userManager.get() != null) {
            if (makeApiCallButton.visibility == View.GONE) {
                makeApiCallButton.visibility = View.VISIBLE
                makeApiCallButton.setOnClickListener { onMakeApiCallClicked() }
            }
            if (signOutButton.visibility == View.GONE) {
                signOutButton.visibility = View.VISIBLE
                signOutButton.setOnClickListener { onSignOutClicked() }
            }

            usernameTextView.text = userManager.get()?.username
            Picasso.with(this)
                    .load(userManager.get()?.avatar)
                    .placeholder(R.drawable.ring_accent)
                    .into(avatarImageView)

        } else {
            makeApiCallButton.visibility = View.GONE
            signOutButton.visibility = View.GONE
        }
    }

    private fun onSignOutClicked() {
        userManager.set(null)
        authService.logout()
    }

    private fun onMakeApiCallClicked() {
        object : AsyncTask<String, Void, String>() {
            override fun doInBackground(vararg tokens: String): String? {
                val client = OkHttpClient()
                val request = Request.Builder()
                        .get()
                        .header("x-auth-token", tokens[0])
                        .url("http://10.0.2.2:5000/api/routines")
                        .build()
                try {
                    val response = client.newCall(request).execute()
                    val jsonBody = response.body()!!.string()
                    return jsonBody
                } catch (exception: Exception) {
                    Log.w(LOG_TAG, exception)
                    return null
                }
            }

            override fun onPostExecute(jsonBody: String?) {
                Toast.makeText(this@MainActivity, jsonBody, Toast.LENGTH_LONG).show();
            }
        }.execute(userManager.get()?.authToken)
    }

    companion object {
        private val LOG_TAG = "MainActivity"
    }
}
